Some tests fail with scikit-learn 0.24

Following #272, some tests fail with scikit-learn 0.24. On Windows 7/Python 3.8.3, I get:
<pre>
FAILED smt/tests/test_all.py::Test::test_tanh_GEKPLS_Cobyla - StopIteration
FAILED smt/tests/test_all.py::Test::test_tanh_GEKPLS_TNC - StopIteration
FAILED smt/tests/test_all.py::Test::test_tanh_KPLSK_Cobyla - AssertionError: 0.010782962850413032 not less than or eq...
FAILED smt/tests/test_all.py::Test::test_tanh_KPLSK_TNC - AssertionError: 0.01067608879639801 not less than or equal ...
FAILED smt/tests/test_all.py::Test::test_tanh_KPLS_Cobyla - AssertionError: 0.13695941639924267 not less than or equa...
FAILED smt/tests/test_all.py::Test::test_tanh_KPLS_TNC - AssertionError: 0.1369637566349206 not less than or equal to...
FAILED smt/tests/test_all.py::Test::test_tanh_MGP - AssertionError: 0.13845802478297414 not less than or equal to 0.0201
================================= 7 failed, 41 passed, 1 warning in 113.46s (0:01:53) =================================
</pre>

It seems to be related to changes to PLS which does not work as well as before with tanh function.

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -27,7 +27,7 @@ jobs:
     - name: Install dependencies
       run: |
         pip install --upgrade pip
-        pip install numpy scipy cython scikit-learn==0.23.2 pytest
+        pip install numpy scipy cython scikit-learn pytest
         pip install git+https://github.com/hwangjt/sphinx_auto_embed.git
         pip install -e .
 

--- a/.github/workflows/tests_coverage.yml
+++ b/.github/workflows/tests_coverage.yml
@@ -27,7 +27,7 @@ jobs:
     - name: Install dependencies
       run: |
         pip install --upgrade pip
-        pip install numpy scipy cython scikit-learn==0.23.2 pytest
+        pip install numpy scipy cython scikit-learn pytest
         pip install git+https://github.com/hwangjt/sphinx_auto_embed.git
         pip list
         pip install -e .

--- a/smt/applications/mfk.py
+++ b/smt/applications/mfk.py
@@ -195,12 +195,20 @@ class MFK(KrgBased):
     def _new_train_init(self):
         if self.name in ["MFKPLS", "MFKPLSK"]:
             _pls = pls(self.options["n_comp"])
-            # PLS is done on the highest fidelity identified by the key None
-            self.m_pls = _pls.fit(
-                self.training_points[None][0][0].copy(),
-                self.training_points[None][0][1].copy(),
-            )
-            self.coeff_pls = self.m_pls.x_rotations_
+
+            # As of sklearn 0.24.1 PLS with zeroed outputs raises an exception while sklearn 0.23 returns zeroed x_rotations
+            # For now the try/except below is a workaround to restore the 0.23 behaviour
+            try:
+                # PLS is done on the highest fidelity identified by the key None
+                self.m_pls = _pls.fit(
+                    self.training_points[None][0][0].copy(),
+                    self.training_points[None][0][1].copy(),
+                )
+                self.coeff_pls = self.m_pls.x_rotations_
+            except StopIteration:
+                self.coeff_pls = np.zeros(
+                    self.training_points[None][0][0].shape[1], self.options["n_comp"]
+                )
 
         xt = []
         yt = []

--- a/smt/applications/mfkpls.py
+++ b/smt/applications/mfkpls.py
@@ -58,7 +58,12 @@ class MFKPLS(MFK):
 
     def _compute_pls(self, X, y):
         _pls = pls(self.options["n_comp"])
-        self.coeff_pls = _pls.fit(X.copy(), y.copy()).x_rotations_
+        # As of sklearn 0.24.1 PLS with zeroed outputs raises an exception while sklearn 0.23 returns zeroed x_rotations
+        # For now the try/except below is a workaround to restore the 0.23 behaviour
+        try:
+            self.coeff_pls = _pls.fit(X.copy(), y.copy()).x_rotations_
+        except StopIteration:
+            self.coeff_pls = np.zeros((X.shape[1], self.options["n_comp"]))
 
         return X, y
 

--- a/smt/surrogate_models/kpls.py
+++ b/smt/surrogate_models/kpls.py
@@ -4,6 +4,8 @@ Author: Dr. Mohamed A. Bouhlel <mbouhlel@umich.edu>
 This package is distributed under New BSD license.
 """
 
+import numpy as np
+
 from packaging import version
 from sklearn import __version__ as sklversion
 
@@ -35,7 +37,12 @@ class KPLS(KrgBased):
 
     def _compute_pls(self, X, y):
         _pls = pls(self.options["n_comp"])
-        self.coeff_pls = _pls.fit(X.copy(), y.copy()).x_rotations_
+        # As of sklearn 0.24.1 zeroed outputs raises an exception while sklearn 0.23 returns zeroed x_rotations
+        # For now the try/except below is a workaround to restore the 0.23 behaviour
+        try:
+            self.coeff_pls = _pls.fit(X.copy(), y.copy()).x_rotations_
+        except StopIteration:
+            self.coeff_pls = np.zeros((X.shape[1], self.options["n_comp"]))
         return X, y
 
     def _componentwise_distance(self, dx, opt=0, theta=None, return_derivative=False):

--- a/smt/surrogate_models/krg_based.py
+++ b/smt/surrogate_models/krg_based.py
@@ -174,11 +174,9 @@ class KrgBased(SurrogateModel):
             self.optimal_noise = np.array(self.options["noise0"])
         elif self.options["use_het_noise"]:
             # hetGP works with unique design variables when noise variance are not given
-            (
-                self.X_norma,
-                index_unique,
-                nt_reps,
-            ) = np.unique(self.X_norma, return_inverse=True, return_counts=True, axis=0)
+            (self.X_norma, index_unique, nt_reps,) = np.unique(
+                self.X_norma, return_inverse=True, return_counts=True, axis=0
+            )
             self.nt = self.X_norma.shape[0]
 
             # computing the mean of the output per unique design variable (see Binois et al., 2018)
@@ -1083,10 +1081,7 @@ class KrgBased(SurrogateModel):
                         [theta0, np.log10(np.array([self.noise0]).flatten())]
                     )
                     theta0_rand = np.concatenate(
-                        [
-                            theta0_rand,
-                            np.log10(np.array([self.noise0]).flatten()),
-                        ]
+                        [theta0_rand, np.log10(np.array([self.noise0]).flatten()),]
                     )
 
                     for i in range(len(self.noise0)):
@@ -1104,14 +1099,14 @@ class KrgBased(SurrogateModel):
                     np.log10([theta_bounds]), repeats=len(theta0), axis=0
                 )
                 theta_all_loops = np.vstack((theta0, theta0_rand))
-                
+
                 if self.options["n_start"] > 1:
                     sampling = LHS(
                         xlimits=theta_limits, criterion="maximin", random_state=41
                     )
                     theta_lhs_loops = sampling(self.options["n_start"])
                     theta_all_loops = np.vstack((theta_all_loops, theta_lhs_loops))
-                    
+
                 optimal_theta_res = {"fun": float("inf")}
                 try:
                     if self.options["hyper_opt"] == "Cobyla":

--- a/smt/surrogate_models/tests/test_kpls.py
+++ b/smt/surrogate_models/tests/test_kpls.py
@@ -39,6 +39,17 @@ class TestKPLS(unittest.TestCase):
         var = kriging.predict_variances(x)
         self.assertEqual(y.shape[0], var.shape[0])
 
+    def test_kpls_training_with_zeroed_outputs(self):
+        # Test scikit-learn 0.24 regression cf. https://github.com/SMTorg/smt/issues/274
+        x = np.random.rand(50, 3)
+        y = np.zeros(50)
+
+        kpls = KPLS()
+        kpls.set_training_values(x, y)
+
+        # KPLS training fails anyway but not due to PLS exception StopIteration
+        self.assertRaises(ValueError, kpls.train)
+
 
 if __name__ == "__main__":
     unittest.main()

--- a/smt/tests/test_all.py
+++ b/smt/tests/test_all.py
@@ -127,7 +127,7 @@ class Test(SMTestCase):
         sname = method_name.split("_")[2]
 
         prob = self.problems[pname]
-        sampling = FullFactorial(xlimits=prob.xlimits, clip=True)
+        sampling = LHS(xlimits=prob.xlimits, random_state=42)
 
         np.random.seed(0)
         xt = sampling(self.nt)

--- a/smt/utils/kriging_utils.py
+++ b/smt/utils/kriging_utils.py
@@ -109,7 +109,7 @@ def cross_distances(X):
 
     n_samples, n_features = X.shape
     n_nonzero_cross_dist = n_samples * (n_samples - 1) // 2
-    ij = np.zeros((n_nonzero_cross_dist, 2), dtype=np.int)
+    ij = np.zeros((n_nonzero_cross_dist, 2), dtype=np.int32)
     D = np.zeros((n_nonzero_cross_dist, n_features))
     ll_1 = 0
 
@@ -120,7 +120,7 @@ def cross_distances(X):
         ij[ll_0:ll_1, 1] = np.arange(k + 1, n_samples)
         D[ll_0:ll_1] = X[k] - X[(k + 1) : n_samples]
 
-    return D, ij.astype(np.int)
+    return D, ij.astype(np.int32)
 
 
 def gower_distances(X, y=None):
@@ -686,10 +686,10 @@ def matern52(theta, d, grad_ind=None, hess_ind=None, derivative_params=None):
                     * fact_2
                 )
                 r[i * nb_limit : (i + 1) * nb_limit, 0] = (
-                    (fact_4 - fact_1 ** 2) / (fact_2) ** 2
-                ) * M52[i * nb_limit : (i + 1) * nb_limit, 0] + r[
-                    i * nb_limit : (i + 1) * nb_limit, 0
-                ]
+                    ((fact_4 - fact_1 ** 2) / (fact_2) ** 2)
+                    * M52[i * nb_limit : (i + 1) * nb_limit, 0]
+                    + r[i * nb_limit : (i + 1) * nb_limit, 0]
+                )
 
             i += 1
     if derivative_params is not None:
@@ -826,13 +826,15 @@ def matern32(theta, d, grad_ind=None, hess_ind=None, derivative_params=None):
             )
             if grad_ind == hess_ind:
                 fact_3 = (
-                    3.0 * d[i * nb_limit : (i + 1) * nb_limit, hess_ind] ** 2.0
-                ) / (
-                    1.0
-                    + np.sqrt(3.0)
-                    * theta_r[0, hess_ind]
-                    * d[i * nb_limit : (i + 1) * nb_limit, hess_ind]
-                ) ** 2.0
+                    (3.0 * d[i * nb_limit : (i + 1) * nb_limit, hess_ind] ** 2.0)
+                    / (
+                        1.0
+                        + np.sqrt(3.0)
+                        * theta_r[0, hess_ind]
+                        * d[i * nb_limit : (i + 1) * nb_limit, hess_ind]
+                    )
+                    ** 2.0
+                )
                 r[i * nb_limit : (i + 1) * nb_limit, 0] = (
                     r[i * nb_limit : (i + 1) * nb_limit, 0]
                     - fact_3 * M32[i * nb_limit : (i + 1) * nb_limit, 0]
@@ -1059,8 +1061,14 @@ def ge_compute_pls(X, y, n_comp, pts, delta_x, xlimits, extra_points):
                 - pts[None][2][1][i, 0] * delta_x * (xlimits[1, 1] - xlimits[1, 0])
             )
 
-        _pls.fit(_X.copy(), _y.copy())
-        coeff_pls[i, :, :] = _pls.x_rotations_
+        # As of sklearn 0.24.1 a zeroed _y raises an exception while sklearn 0.23 returns zeroed x_rotations
+        # For now the try/except below is a workaround to restore the 0.23 behaviour
+        try:
+            _pls.fit(_X.copy(), _y.copy())
+            coeff_pls[i, :, :] = _pls.x_rotations_
+        except StopIteration:
+            coeff_pls[i, :, :] = 0
+
         # Add additional points
         if extra_points != 0:
             max_coeff = np.argsort(np.abs(coeff_pls[i, :, 0]))[-extra_points:]

